File: shared/src/commonMain/kotlin/com/phlick/game/RequiredPrayers.kt
package com.phlick.game

/**
 * Returns the set of prayers required to block all attacks on the current tick.
 * Used for attack resolution and UI.
 */
fun computeRequiredPrayers(
    gameMode: GameMode,
    currentTick: Int,
    currentLevel: Level?,
    reactiveMonsterAttacks: Map<Int, Prayer>,
    magicMonsterAttackOffset: Int?,
    rangedMonsterAttackOffset: Int?
): Set<Prayer> = when (gameMode) {
    GameMode.Progression -> {
        val effectiveTick = currentTick - Levels.PROGRESSION_WARMUP_TICKS
        if (effectiveTick < 0) emptySet()
        else {
            currentLevel?.monsters?.mapNotNull { monster ->
                val offset = monster.offset
                val adjustedTick = currentTick - offset
                if (adjustedTick > 0 && adjustedTick % monster.cycleLength == 0 && effectiveTick >= offset) {
                    if (monster.isReactive) {
                        val monsterIndex = currentLevel.monsters.indexOf(monster)
                        reactiveMonsterAttacks[monsterIndex] ?: monster.attackStyle
                    } else {
                        monster.attackStyle
                    }
                } else null
            }?.toSet() ?: emptySet()
        }
    }
    GameMode.Sandbox -> {
        val magicAttack = magicMonsterAttackOffset?.let { offset ->
            (currentTick - offset) % 4 == 0 && currentTick >= offset
        } ?: false
        val rangedAttack = rangedMonsterAttackOffset?.let { offset ->
            (currentTick - offset) % 4 == 0 && currentTick >= offset
        } ?: false
        buildSet {
            if (magicAttack) add(Prayer.Magic)
            if (rangedAttack) add(Prayer.Missiles)
        }
    }
}

File: web/src/core/requiredPrayers.ts
import type { GameMode, Level, Prayer } from "./types";
import { PROGRESSION_WARMUP_TICKS } from "./constants";

export function computeRequiredPrayers(
  gameMode: GameMode,
  currentTick: number,
  currentLevel: Level | null,
  reactiveMonsterAttacks: Record<number, Prayer>,
  magicMonsterAttackOffset: number | null,
  rangedMonsterAttackOffset: number | null
): Set<Prayer> {
  if (gameMode === "Progression") {
    const effectiveTick = currentTick - PROGRESSION_WARMUP_TICKS;
    if (effectiveTick < 0) return new Set();
    if (!currentLevel) return new Set();
    const required = new Set<Prayer>();
    currentLevel.monsters.forEach((monster, index) => {
      const offset = monster.offset;
      const adjustedTick = currentTick - offset;
      if (adjustedTick > 0 && adjustedTick % monster.cycleLength === 0 && effectiveTick >= offset) {
        const style = monster.isReactive
          ? reactiveMonsterAttacks[index] ?? monster.attackStyle
          : monster.attackStyle;
        required.add(style);
      }
    });
    return required;
  }
  // Sandbox
  const magicAttack =
    magicMonsterAttackOffset != null &&
    (currentTick - magicMonsterAttackOffset) % 4 === 0 &&
    currentTick >= magicMonsterAttackOffset;
  const rangedAttack =
    rangedMonsterAttackOffset != null &&
    (currentTick - rangedMonsterAttackOffset) % 4 === 0 &&
    currentTick >= rangedMonsterAttackOffset;
  const set = new Set<Prayer>();
  if (magicAttack) set.add("Magic");
  if (rangedAttack) set.add("Missiles");
  return set;
}
